index/scorch/persister.go

```diff
@@ -149,11 +152,25 @@ OUTER:
 					_ = ourSnapshot.DecRef()
 					break OUTER
 				}
+
+				// save this current snapshot's persistedCallbacks, to invoke during
+				// the retry attempt
+				unpersistedCallbacks = append(unpersistedCallbacks, ourPersistedCallbacks...)
+
 				s.fireAsyncError(fmt.Errorf("got err persisting snapshot: %v", err))
 				_ = ourSnapshot.DecRef()
 				atomic.AddUint64(&s.stats.TotPersistLoopErr, 1)
 				continue OUTER
 			}
+
+			if unpersistedCallbacks != nil {
```

Comment:
shouldn't we need an  " unpersistedEpoch <= ourSnapshot.Epoch " check as well to guard against cases of rollbacks where the cached callbacks (unpersistedCallbacks) becomes irrelevant?

Reply:
Hmm, I think we can do without, as along as there are unpersisted callbacks we should invoke. For even in case of a rollback, a new snapshot of segments will be introduced with a monotonically increasing epoch.

Reply:
you are right on the epoch numbers, But is it the right thing to fire a callback when things rollbacked and becomes irrelevant?
May be thats a CB specific concern, not a bleve specific one.

Reply:
I guess I'm not seeing what the concern is, could you elaborate a bit on it?
Also, if it's the couchbase case only - a fireAsyncError is treated as Fatal :)

Reply:
The concern was that - in case of situations like a rollback - the most recent/last snapshots become irrelevant and thus the associated batch contents/callbacks. So in such cases, we don't have to care about appending these cached callbacks and is there a way to identify or handle such cases.